src/hypercorn/asyncio/tcp_server.py

```diff
@@ -102,10 +104,19 @@ async def _read_data(self) -> None:
                 SSLError,
             ):
                 await self.protocol.handle(Closed())
+                closed_sent = True
                 break
             else:
                 await self.protocol.handle(RawData(data))
 
+        # It happens with TLS connections that we don't got a
+        # `ConnectionError`, but the connection gets closed at some point. We
+        # want to make sure that a `Closed()` got fed into the protocol, so
+        # that protocol handlers will be unblocked.
+        # See: https://github.com/pgjones/hypercorn/issues/50
+        if not closed_sent:
```

Comment:
Code-wise, this is only true if the `sent=True;break` terminated the loop.

Thus, I think this alternative coding style would do the same, perhaps more Pythonic, but maybe somewhere less explicit:

```py
while not eof():
    under some cond:
        await handle(closed())
        break
else:
    await handle(Closed())
```